src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        [ThreadStatic] private static DisposeScopeManager _threadSingleton;
+
+        internal static DisposeScopeManager ThreadSingleton => (_threadSingleton ??= new DisposeScopeManager());
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        public long CreatedOutsideScopeCount { get; private set; }
```

Comment:
Should these properties be collected in a "Statistics" class? I don't have a strong opinion, just bringing it up.

Reply:
Could do, I'll leave it here for now and see what else crops up, I love statistics, it really helps debugging and optimizing, so if we add more stuff it might be a good idea.

Reply:
I did this now, it simplified a couple of other things;

    /// <summary>
    /// Keeps track of statistics for a dispose scope. Can be queried to figure out performance/memory issues.
    /// </summary>
    public class ThreadDisposeScopeStatistics
    {
        /// <summary>
        /// The number of disposables that were created on this thread, but weren't captured by a DisposeScope.
        /// </summary>
        public long CreatedOutsideScopeCount { get; internal set; }

        /// <summary>
        /// The number of disposables that were created on this thread and were captured by a DisposeScope.
        /// </summary>
        public long CreatedInScopeCount { get; internal set; }

        /// <summary>
        /// The number of disposables that were disposed on this thread and were disposed while in a DisposeScope.
        /// </summary>
        public long DisposedInScopeCount { get; internal set; }

        /// <summary>
        /// Number of disposables that were once included in the scope, but were subsequently detached.
        /// </summary>
        public long DetachedFromScopeCount { get; internal set; }

        /// <summary>
        /// The number of disposables that are currently live on the current thread. If a It's aproximate, see
        /// Tensor.TotalCount. Disposables that weren't created within a DisposeScope, or detached from the dispose
        /// scope, will not be counted as alive.
        /// </summary>
        public long ThreadTotalLiveCount => CreatedInScopeCount - DisposedInScopeCount - DetachedFromScopeCount;

        /// <summary>
        /// Resets the counts for the current thread. See ThreadTotalLiveCount etc. Mainly used in tests to make sure
        /// we get a clean slate on the thread.
        /// </summary>
        public void Reset()
        {
            CreatedOutsideScopeCount = 0;
            CreatedInScopeCount = 0;
            DisposedInScopeCount = 0;
            DetachedFromScopeCount = 0;
        }
    }